### DISCUSSION TITLE
How does array embedding work in mruby?

### DISCUSSION THREAD

**Discussion by sasq64** [mruby/mruby#5472] [Q&A]
How does array embedding (and the ARY_EMBED_ macros) work in mruby ?

I am using a lot of 2 element float arrays in my code (points) and want to make sure they are fast, and handled quickly on
the C-side.

**Comment by matz** [comment]
If the number of array elements is smaller than `MRB_ARY_EMBED_LEN_MAX`, they are automatically embedded in the `RVALUE` struct.
The actual value of `MRB_ARY_EMBED_LEN_MAX` is configuration dependant. If you are using `MRB_WORD_BOXING` (which is default), it should be `3`.

**Reply by sasq64** [reply]
But are they embedded as mrb_values ? or as the underlying type ?

If I know for instance that I have an array of exactly 2 floats, what is the easiest way to access those floats from the C-side ?

**Reply by matz** [reply]
`mrb_value` (its minimum size is a word) is too small to embed array objects. Mutable objects like arrays cannot be embedded in `mrb_value` anyway. `EMBED` in `array.c` and `string.c` mean embedding in `RVALUE` not `mrb_value`.

Values embeddable  in `mrb_value` are:

* Integers
* Symbols
* Floats (if `MRB_NAN_BOXING`)
